### PR TITLE
feat(crypto): Add signalling to the SAS verification

### PR DESCRIPTION
**Pull request by poljar**
This patch adds a way for users to listen to changes in the state of a SAS verification.

This makes it much more pleasant to go through the verification flow and incidentally easier to document it.

### CODE REVIEW

File: crates/matrix-sdk-crypto/src/verification/sas/sas_state.rs [hunk 1]
```diff
@@ -239,21 +240,22 @@ impl<S: Clone + std::fmt::Debug> std::fmt::Debug for SasState<S> {
 /// The initial SAS state.
 #[derive(Clone, Debug)]
 pub struct Created {
-    protocol_definitions: SasV1Content,
+    pub protocol_definitions: SasV1Content,
```
Comment (gnunicorn) [inline crates/matrix-sdk-crypto/src/verification/sas/sas_state.rs:243]: isn't the compiler/clippy complaining that you made this public but don't give doc comments for it (also the ones below)? Not that I think it makes a lot of sense here, it is pretty clear naming ... just curious...

File: crates/matrix-sdk/src/encryption/verification/sas.rs [hunk 2]
```diff
@@ -217,4 +220,108 @@ impl SasVerification {
     pub fn other_user_id(&self) -> &UserId {
         self.inner.other_user_id()
     }
+
+    /// Listen for changes in the SAS verification process.
```
Comment (gnunicorn) [inline crates/matrix-sdk/src/encryption/verification/sas.rs:224]: :heavy_plus_sign: for the docs!

**Review by gnunicorn — APPROVED** [review]
I am not deeply familiar with the process itself, so there might be a few blind spots on edge cases I wouldn't have noticed, but it looks fine with me overall. Just some minor questions.

File: crates/matrix-sdk-crypto/src/verification/sas/sas_state.rs (hunk 1 above)
Comment (poljar) [reply crates/matrix-sdk-crypto/src/verification/sas/sas_state.rs:243]: The struct and field are public in the module and in the parent module, but it's never exposed further from the parent module.

**Pull request merged**